akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1927,3 +1928,152 @@ private[stream] object Collect {
   override def toString = "StatefulMapConcat"
 
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi final private[akka] class LazyFlow[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]], zeroMat: () ⇒ M)
+  extends GraphStageWithMaterializedValue[FlowShape[I, O], M] {
+  val in = Inlet[I]("lazyFlow.in")
+  val out = Outlet[O]("lazyFlow.out")
+  override def initialAttributes = DefaultAttributes.lazyFlow
+  override val shape: FlowShape[I, O] = FlowShape.of(in, out)
+
+  override def toString: String = "LazyFlow"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+    var completed = false
+    var matVal: Option[M] = None
+    val stageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      val subSink = new SubSinkInlet[O]("LazyFlowSubSink")
+      override def preStart(): Unit = pull(in)
+
+      override def onPush(): Unit = {
+        try {
+          val element = grab(in)
+          val cb: AsyncCallback[Try[Flow[I, O, M]]] =
+            getAsyncCallback {
+              case Success(flow) ⇒ initInternalSource(flow, element)
+              case Failure(e)    ⇒ failure(e)
+            }
+          flowFactory(element).onComplete { cb.invoke }(ExecutionContexts.sameThreadExecutionContext)
+          setHandler(in, new InHandler {
+            override def onPush(): Unit = ()
+            override def onUpstreamFinish(): Unit = gotCompletionEvent()
+            override def onUpstreamFailure(ex: Throwable): Unit = failure(ex)
+          })
+        } catch {
+          case NonFatal(e) ⇒ decider(e) match {
+            case Supervision.Stop ⇒ failure(e)
+            case _                ⇒ pull(in)
+          }
+        }
+      }
+
+      override def onPull(): Unit = {
+        subSink.pull()
```

Comment:
Regarding the pull(in) from `preStart`. Wouldn't this be the place to do that instead? It's replacing the OutHandler immediately so it will only be done once on first pull from downstream? What am I missing?

Reply:
Unfortunately, it seems this is a bit too lazy. It would fail all tests, where we do not request an element in downstream (e.g. "must fail gracefully when flow factory method failed", "must cancel upstream when the downstream is cancelled" or " must continue if supervision is resume").

Question - are there any valid reasons for such scenarios to work, or shall I also tweak tests to generate some demand from downstream?

Reply:
Sounds like such tests can request early from downstream to trigger the initialization.
I think it's most intuitive that it's as lazy as possible, instead of half lazy.